### PR TITLE
Do not allocate memory when there's no constraints

### LINKED ISSUE #1276: constrainLabels allocates too many memory
Hi, in my application prometheus allocates a lot of memory in constrainLabels.
I suggest replacing labels inplace instead of creating a new map.
image

### PR DESCRIPTION
**Pull request by Okhoshi**
Implementing a fast path when there's no constrained label should limit a lot the number of allocations.

Closes #1276

### CODE REVIEW

File: prometheus/benchmark_test.go [hunk 1]
```diff
@@ -56,6 +79,37 @@ func BenchmarkCounterWithLabelValuesConcurrent(b *testing.B) {
 	wg.Wait()
 }
 
+func BenchmarkCounterWithLabelValuesAndConstraintsConcurrent(b *testing.B) {
```
Comment (bwplotka) [inline prometheus/benchmark_test.go:82]: Those feel duplicated below, no? Or at least the difference is small.

Comment (Okhoshi) [reply prometheus/benchmark_test.go:82]: It's indeed a small difference, there's a constraint whereas the first version doesn't have any constraint at all. It lets us assess the impact of constraints on performances and allocations.

**Review by bwplotka — COMMENTED** [review]
Nice, thanks for this!

I think it makes sense, but we need to clean up those benchmarks. They are mostly duplicate and will get obsolete very soon. So either we use b.Run or remove them for now.

BTW Were those useful? Did you gather some info? Or just helped you to finalized the implementation? (:

File: prometheus/desc.go [hunk 2]
```diff
@@ -189,11 +189,19 @@ func (d *Desc) String() string {
 			fmt.Sprintf("%s=%q", lp.GetName(), lp.GetValue()),
 		)
 	}
+	vlStrings := make([]string, 0, len(d.variableLabels.names))
+	for _, vl := range d.variableLabels.names {
+		if fn, ok := d.variableLabels.labelConstraints[vl]; ok && fn != nil {
+			vlStrings = append(vlStrings, fmt.Sprintf("constrained(%s)", vl))
+		} else {
+			vlStrings = append(vlStrings, fmt.Sprintf("unconstrained(%s)", vl))
```
[the remark this answers is not in the document]
Comment (Okhoshi) [reply prometheus/desc.go:197]: Makes sense 👍

File: prometheus/labels.go [hunk 3]
```diff
@@ -92,18 +100,37 @@ func (cls ConstrainedLabels) labelNames() []string {
...
 type UnconstrainedLabels []string
 
-func (uls UnconstrainedLabels) constrainedLabels() ConstrainedLabels {
-	constrainedLabels := make([]ConstrainedLabel, len(uls))
-	for i, l := range uls {
-		constrainedLabels[i] = ConstrainedLabel{Name: l}
+func (uls UnconstrainedLabels) compile() *compiledLabels {
+	return &compiledLabels{
+		names:            uls,
+		labelConstraints: map[string]LabelConstraint{},
```
[the remark this answers is not in the document]
Comment (Okhoshi) [reply prometheus/labels.go:106]: I was also considering that option, will do the change 👍

**Comment by Okhoshi** [comment]
> BTW Were those useful? Did you gather some info? Or just helped you to finalized the implementation? (:

They essentially helped me to see the progress and if there were some remaining optimisations that could be done.

**Comment by bwplotka** [comment]
Great 💪🏽 

I feel we can either drop those benchmark for now if you don't have time to clean them or create one or two with table tests so others can run it on their PRs to verify if there is any regression 🤗

**Comment by Okhoshi** [comment]
Sorry for the delay, I got busy at work and couldn't find the time before :) I rewrote the benchmarks as a table test, it should be easier to maintain like this 👍

**Review by bwplotka — APPROVED** [review]
So much better, thanks so much 💪🏽

**Comment by marten-seemann** [comment]
Any chance we cut get a release that includes this fix?

We have some tests that our metrics collection doesn't allocate in [go-libp2p](https://github.com/libp2p/go-libp2p), and we currently can't upgrade this dependency past v1.14.0 without causing these tests to fail.
Obviously, no matter which version we specify in go-libp2p's go.mod, users of our library might still have client_golang bumped to a version greater than v1.14.0, causing a performance regression.

**Pull request merged**